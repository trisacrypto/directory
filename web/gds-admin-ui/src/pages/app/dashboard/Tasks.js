import React from 'react';
import { Card, Dropdown, Table } from 'react-bootstrap';
import { useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom'
import dayjs from 'dayjs';
import ResendEmail from '../../../components/ResendEmail';
import { StatusLabel } from '../../../constants';
import TrisaFavicon from '../../../assets/images/trisa_favicon.png'
import CiphertraceFavicon from '../../../assets/images/ciphertrace.ico'

import relativeTime from 'dayjs/plugin/relativeTime'
dayjs.extend(relativeTime)




const Tasks = () => {
    const { vasps } = useSelector((state) => ({
        vasps: state.Vasps.data,
        certificates: state.Certificates.data
    }));
    const [modal, setModal] = React.useState(false);
    const [vaspName, setVaspName] = React.useState("");
    const history = useHistory()

    const toggle = () => {
        setModal(!modal);
    };

    const handleResendEmailClick = (name) => {
        setVaspName(name)
        toggle()
    }


    return (
        <Card>
            <Card.Body>
                <h4 className="header-title mb-3">Pending Reviews</h4>

                <Table responsive className="table table-centered table-nowrap table-hover mb-0 z-index-2">
                    <tbody>
                        {
                            Array.isArray(vasps?.vasps) && vasps?.vasps.map((vasp) => vasp?.verification_status === "PENDING_REVIEW" && (

                                <tr key={vasp.id}>
                                    <td className="d-flex gap-2 align-items-center">
                                        <div>
                                            {
                                                vasp?.is_traveler ? <img src={CiphertraceFavicon} className="img-fluid" alt="Cyphertrace" /> : <img src={TrisaFavicon} className="img-fluid" alt="Trisa" />
                                            }
                                        </div>
                                        <div>
                                            <h5 className="font-14 my-1 d-flex gap-2">
                                                <a href="/" className="text-body">
                                                    {vasp?.name}
                                                </a>
                                            </h5>
                                            <span className="text-muted font-13">{dayjs(vasp?.last_updated).fromNow()}</span>
                                        </div>
                                    </td>
                                    <td>
                                        <span className="text-muted font-13">Status</span> <br />
                                        <span className="badge badge-warning-lighten">{StatusLabel[vasp?.verification_status]}</span>
                                    </td>
                                    <td>
                                        <span className="text-muted font-13">Emails</span>
                                        <div className="font-14 mt-1 fw-normal">
                                            {
                                                vasp?.verified_contacts && Object.keys(vasp?.verified_contacts).map(contact => (
                                                    <span key={contact} className={`badge ${vasp?.verified_contacts[contact] ? "badge-success-lighten" : "badge-danger-lighten"}`} style={{ marginRight: 4 }}>{contact}</span>
                                                )
                                                )
                                            }
                                        </div>
                                    </td>
                                    <td className="table-action text-center" style={{ width: '90px' }}>
                                        <Dropdown align="end">
                                            <Dropdown.Toggle variant="link" className="arrow-none card-drop p-0 shadow-none">
                                                <i className="mdi mdi-dots-horizontal"></i>
                                            </Dropdown.Toggle>
                                            <Dropdown.Menu>
                                                <Dropdown.Item onClick={() => history.push(`/vasps/${vasp?.id}/details`)}> <span className="mdi mdi-eye-outline"></span> View</Dropdown.Item>
                                                <Dropdown.Item onClick={() => handleResendEmailClick(vasp?.name)}> <span className="mdi mdi-email-edit-outline"></span> Resend email</Dropdown.Item>
                                            </Dropdown.Menu>
                                        </Dropdown>
                                    </td>
                                </tr>
                            ))
                        }
                    </tbody>
                </Table>
                <ResendEmail toggle={toggle} modal={modal} vaspName={vaspName} />
            </Card.Body>
        </Card>
    );
};

export default Tasks;
