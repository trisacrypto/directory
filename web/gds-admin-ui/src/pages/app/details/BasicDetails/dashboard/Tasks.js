/* eslint-disable no-unused-vars */
import React from 'react';
import { Card, Dropdown, Table } from 'react-bootstrap';
import { useHistory } from 'react-router-dom'
import dayjs from 'dayjs';
import { StatusLabel, Status as VerificationStatus } from 'constants/index';
import TrisaFavicon from 'assets/images/trisa_favicon.png'
import CiphertraceFavicon from 'assets/images/ciphertrace.ico'
import SimpleBar from 'simplebar-react';
import PropTypes from 'prop-types';
import relativeTime from 'dayjs/plugin/relativeTime'
import { actionType, useModal } from 'contexts/modal';
import OvalLoader from 'components/OvalLoader';
import NoData from 'components/NoData';
import filter from 'lodash/filter';
import isDate from 'lodash/isDate';
dayjs.extend(relativeTime)

/**
 * Verifiy that the passed date is less than 30days
 * @param {date} date 
 * @returns 
 */
function isRecent(date = '') {

    if (dayjs(date).isValid) {
        const tirthyDaysInMs = 30 * 24 * 60 * 60 * 1000;
        const now = dayjs();
        const dateDiff = now.diff(date)

        return Math.abs(dateDiff) <= tirthyDaysInMs
    }

    return false
}

export const getRecentVasps = (vasps) => vasps?.filter(vasp => {
    const now = dayjs()
    const certificateExpirationDate = vasp?.certificate_expiration && dayjs(vasp?.certificate_expiration)
    const certificateIssuedDate = vasp?.certificate_issued && dayjs(vasp?.certificate_issued)
    const verificationStatus = vasp?.verification_status

    if (verificationStatus === VerificationStatus.PENDING_REVIEW) {
        return vasp
    }
    if (certificateExpirationDate && certificateIssuedDate) {

        if (dayjs(certificateExpirationDate).isValid() || dayjs(certificateIssuedDate).isValid()) {

            if (certificateExpirationDate.isAfter(now) && isRecent(certificateExpirationDate)) {
                return vasp
            }

            if (certificateIssuedDate.isBefore(now) && isRecent(certificateIssuedDate)) {
                return vasp
            }
        }
    }
})

const PendingReviewsTable = ({ data }) => {
    const [vasp, setVasp] = React.useState({ name: '', id: '' });
    const { dispatch } = useModal()
    const history = useHistory()

    const handleResendEmailClick = (name) => {
        setVasp(name)
        dispatch({ type: actionType.SEND_EMAIL_MODAL, payload: { vasp } })
    }

    return (
        <SimpleBar style={{ maxHeight: 350 }} className="task">
            <Table responsive className="table table-centered table-nowrap table-hover mb-0 z-index-2">
                <tbody>
                    {
                        data?.map(vasp => (
                            <tr key={vasp.id}>
                                <td onClick={() => history.push(`/vasps/${vasp?.id}`)} className="d-flex gap-2 align-items-center" role="button">
                                    <div>
                                        {
                                            vasp?.traveler ? <img src={CiphertraceFavicon} width="30" alt="CipherTrace" /> : <img src={TrisaFavicon} width="30" className="img-fluid" alt="TRISA" />
                                        }
                                    </div>
                                    <div>
                                        <h5 className="font-14 my-1 gap-2 d-flex">
                                            {vasp?.name}
                                        </h5>
                                        <span className="text-muted font-13">{dayjs(vasp?.last_updated).fromNow()}</span>
                                    </div>
                                </td>
                                <td>
                                    <span className="text-muted font-13">Status</span> <br />
                                    <span className="badge badge-warning-lighten">{StatusLabel[vasp?.verification_status]}</span>
                                </td>
                                <td>
                                    <span className="text-muted font-13">Emails</span>
                                    <div className="font-14 mt-1 fw-normal">
                                        {
                                            vasp?.verified_contacts && Object.keys(vasp?.verified_contacts).map(contact => (
                                                <span key={contact} className={`badge ${vasp?.verified_contacts[contact] ? "badge-success-lighten" : "badge-danger-lighten"}`} style={{ marginRight: 4 }}>{contact}</span>
                                            )
                                            )
                                        }
                                    </div>
                                </td>
                                <td className="table-action text-center" style={{ width: '90px' }}>
                                    <Dropdown align="end">
                                        <Dropdown.Toggle variant="link" className="arrow-none card-drop p-0 shadow-none">
                                            <i className="mdi mdi-dots-horizontal"></i>
                                        </Dropdown.Toggle>
                                        <Dropdown.Menu>
                                            <Dropdown.Item onClick={() => history.push(`/vasps/${vasp?.id}`)}> <span className="mdi mdi-eye-outline"></span> View</Dropdown.Item>
                                            <Dropdown.Item onClick={() => handleResendEmailClick(vasp)}> <span className="mdi mdi-email-edit-outline"></span> Resend email</Dropdown.Item>
                                        </Dropdown.Menu>
                                    </Dropdown>
                                </td>
                            </tr>)
                        )
                    }
                </tbody>
            </Table>
        </SimpleBar>
    )
}

const Tasks = ({ data, isLoading }) => {
    const recentVasps = getRecentVasps(data?.vasps || [])

    if (isLoading || !data) {
        return (
            <Card className='w-100'>
                <Card.Body>
                    <h4 className="header-title mb-3">Pending and Recent Activity</h4>
                    <OvalLoader />
                </Card.Body>
            </Card>
        )
    }

    return (
        <Card className='w-100'>
            <Card.Body>
                <h4 className="header-title mb-3">Pending and Recent Activity</h4>
                {
                    (isLoading || !recentVasps) && <OvalLoader />
                }
                {
                    recentVasps?.length ? <PendingReviewsTable data={recentVasps} /> : <NoData title='No available pending registrations' />
                }
            </Card.Body>
        </Card>
    );
};

Tasks.propTypes = {
    data: PropTypes.object
}

export default Tasks;
