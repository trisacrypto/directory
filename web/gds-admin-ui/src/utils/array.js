// @flow
const groupByFields = (array: Array<any>, f: any): Array<any> => {
  /*
    params description :
        f : function which returnf the array of fields
        e.g. :  (item) => {
            return [itemField1, itemField2];
        }
        array : array of data to group e.g. : [{...}, {...}]
    */
  const groups = {};
  array.forEach((o) => {
    const group = JSON.stringify(f(o));
    groups[group] = groups[group] || [];
    groups[group].push(o);
  });
  return Object.keys(groups).map((group) => groups[group]);
};

export { groupByFields };
