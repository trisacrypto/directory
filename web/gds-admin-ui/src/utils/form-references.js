const normalizeFlatArrays = (data) => data && data.map(d => ({ name: d }))

export const getTrixoFormInitialValues = (data = []) => {
    const defaultValues = {
        applicable_regulations: [{
            name: "FATF Recommendation 16"
        }],
        compliance_threshold: 0,
        compliance_threshold_currency: "",
        conducts_customer_kyc: false,
        financial_transfers_permitted: "",
        has_required_regulatory_program: "",
        kyc_threshold: 0,
        kyc_threshold_currency: "",
        must_comply_travel_rule: false,
        must_safeguard_pii: false,
        other_jurisdictions: [{
            country: "",
            regulator_name: ""
        }],
        primary_national_jurisdiction: "",
        primary_regulator: "",
        safeguards_pii: false
    }

    const applicableRegulations = normalizeFlatArrays(data.applicable_regulations)

    const result = {
        ...Object.assign(defaultValues, data),
        applicable_regulations: applicableRegulations,
    }

    return result
}

export const getBusinessInfosFormInitialValues = (data) => ({
    website: data.vasp.website || "",
    established_on: data.vasp.established_on || "",
    vasp_categories: data.vasp.vasp_categories || [],
    business_category: data.vasp.business_category || ""
})

export const getTrisaImplementationDetailsInitialValue = (data) => ({
    common_name: data.vasp.common_name,
    trisa_endpoint: data.vasp.trisa_endpoint
})

export const getIvms101RecordInitialValues = (data) => {
    const defaultValues = {
        name: {
            name_identifiers: [
                {
                    legal_person_name: "",
                    legal_person_name_identifier_type: "0"
                }
            ],
            local_name_identifiers: [],
            phonetic_name_identifiers: []
        },
        geographic_addresses: [
            {
                address_type: 2,
                address_line: [
                    "",
                    "",
                    ""
                ],
                country: ""
            }
        ],
        customer_number: "",
        national_identification: {
            national_identifier: "",
            national_identifier_type: 0,
            country_of_issue: "",
            registration_authority: ""
        },
        country_of_registration: ""
    }

    return {
        ...Object.assign(defaultValues, data),
    }
}

export const getContactInitialValues = (data) => {
    const initialValue = {
        email: "",
        extra: null,
        name: "",
        person: null,
        phone: ""
    }

    return Object.assign(initialValue, data)
}