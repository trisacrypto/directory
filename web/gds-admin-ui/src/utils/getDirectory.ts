import { DIRECTORY_NAME } from "@/constants";
import { isTestNet } from ".";

export default function getDirectory(){
    return isTestNet() ? DIRECTORY_NAME.TRISATEST : DIRECTORY_NAME.VASP_DIRECTORY
}