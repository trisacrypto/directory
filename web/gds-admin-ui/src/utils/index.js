import config from '../config';
import { ENVIRONMENT } from '../constants';

export * from './array';


function defaultEndpointPrefix() {
    if (config.GDS_API_URL) {
        return config.GDS_API_URL
    }

    switch (process.env.NODE_ENV) {
        case ENVIRONMENT.DEV:
            return "http://localhost:4434/v2"
        case ENVIRONMENT.PROD:
            if (config.IS_TESTNET) {
                return "https://api.admin.trisatest.net/v2";
            } else {
                return "https://api.admin.vaspdirectory.net/v2";
            }
        default:
            throw new Error("Could not identify the api prefix");
    }
}

function apiHost() {
    const url = new URL(config.GDS_API_URL)
    return url.hostname;
}

function formatDisplayedData(target) {
    if (typeof target === "boolean") {
        return target.toString()
    } else if (Array.isArray(target)) {
        return target.length ? target.toString() : "N/A"
    } else if (typeof target === "string") {
        return target ? target.trim() : "N/A"
    }

    return target ? target : "N/A"
}


export { formatDisplayedData, defaultEndpointPrefix, apiHost }