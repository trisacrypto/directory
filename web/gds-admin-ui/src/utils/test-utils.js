import { Provider } from "react-redux";
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from "../redux/store";

export const render = (ui) => rtlRender(<Provider store={configureStore({})}>{ui}</Provider>)

export * from '@testing-library/react'