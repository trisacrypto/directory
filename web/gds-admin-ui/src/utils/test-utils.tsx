import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render as rtlRender } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Suspense } from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';

import { configureStore } from '../redux/store';
import { ModalProvider } from '../contexts/modal';
import React from 'react';

const queryClient = new QueryClient();
export const render = (ui: any) => {
    const history = createMemoryHistory();

    return rtlRender(
        <Provider store={configureStore({})}>
            <QueryClientProvider client={queryClient}>
                <Suspense fallback="loading...">
                    <Router history={history}>
                        <ModalProvider>{ui}</ModalProvider>
                    </Router>
                </Suspense>
            </QueryClientProvider>
        </Provider>
    );
};

export * from '@testing-library/react';
