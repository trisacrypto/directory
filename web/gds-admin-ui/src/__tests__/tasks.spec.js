import Tasks from "pages/app/dashboard/Tasks"
import { render } from "utils/test-utils"

describe("<Tasks />", () => {

    it("Should keep the same UI", () => {
        const { container } = render(<Tasks />)

        //
        expect(container).toMatchSnapshot("pending-registrations")
    })
})