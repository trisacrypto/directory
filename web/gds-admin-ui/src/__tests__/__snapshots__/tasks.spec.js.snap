// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tasks /> Should keep the same UI: pending-registrations 1`] = `
<div>
  <div
    class="w-100 card"
  >
    <div
      class="card-body"
    >
      <h4
        class="header-title mb-3"
      >
        Pending Reviews
      </h4>
      <div
        class="text-center flex flex-column justify-content-center"
      >
        <div>
          <svg
            height="40"
            stroke="#6b7280"
            viewBox="0 0 38 38"
            width="40"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <g
                stroke-width="2"
                transform="translate(1 1)"
              >
                <circle
                  cx="18"
                  cy="18"
                  r="18"
                  stroke-opacity=".5"
                />
                <path
                  d="M36 18c0-9.94-8.06-18-18-18"
                >
                  <animatetransform
                    attributeName="transform"
                    dur=".7s"
                    from="0 18 18"
                    repeatCount="indefinite"
                    to="360 18 18"
                    type="rotate"
                  />
                </path>
              </g>
            </g>
          </svg>
        </div>
        <div
          class="mt-1"
        >
          <small
            class="block"
          >
            Loading...
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;
